File: structy/02-linked-list/05-get-node-value/01.js
/*
reverse list
Write a function, reverseList, that takes in the head of a linked list as an argument. The function should reverse the order of the nodes in the linked list in-place and return the new head of the reversed linked list.
*/

/*
INPUT: Node (head)
OUTPUT: Node (reversed linked list head)

RULES 
- it should reverse the order of the nodes in the linked list 
- it should reverse the linked list "in-place"
- it should return the head of the reverse linked list 

Assumptions 
- A valid linked list with at least one Node will always be provided
    - we'll never get NULL as a value 

Exploration 
T1
tail = null
rest
tail <- a <- b <- c -> d -> e -> f -> NULL

tail = null  

while current !== null 
    current.next = tail
    rest = current.next
    current = current.next 

b -> c -> d -> e -> f -> NULL

a <- b <- c <- d <- e <- f <- NULL

DATA STRUCTURE 
- original linked list (mutated)

ALGORITHM 
--> given a `head` parameter representing a linked list node

- set `prev` to NULL 
- set `curr` to `head` 
- set `next` to `head.next` 

- while `curr` is NOT NULL 
    - set `curr.next` to `prev`
    - set `next` to `curr.next`
    - set `curr` to `next`
    - set `prev` to `curr`
- return `prev`

COMPLEXITY (Iterative solution)
- N = # of nodes
- Time: O(N)
- Space: O(1)
*/

import Node from '../node.js';

/*
const reverseList = (head) => {
    let prev = null; 
    let curr = head; 
    let next = curr.next;

    while (curr !== null) {
        curr.next = prev;
        next = curr.next;
        curr = next; 
        prev = curr;
    }

    return prev;
};
*/

const reverseLinkedList = (head) => {
    let current = head; 
    let prev = null; 
    
    while (current !== null) {
        const next = current.next;
        current.next = prev; 
        prev = current;
        current = next;
    }

    return prev;
}; 


const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.next = b;
b.next = c;
c.next = d;
d.next = e;
e.next = f;

// a -> b -> c -> d -> e -> f

console.log(reverseList(a)); // f -> e -> d -> c -> b -> a


const x = new Node("x");
const y = new Node("y");

x.next = y;

// x -> y

console.log(reverseList(x)); // y -> 


const p = new Node("p");

// p

console.log(reverseList(p)); // p

File: structy/02-linked-list/13-create-linked-list/01.py
# create linked list

# Write a function, create_linked_list, that takes in a list of values as an argument. 
# The function should create a linked list containing each item of the list as the values of the nodes. 
# The function should return the head of the linked list.

"""
INPUT: 
- values: List[Any]

OUTPUT: 
- linked_list: Node

RULES 
Tasks
- it should create a linked list containing each item of the input [regular] list of values 
- it should return the head of the linked list 
- it should return None if the list is empty 

Metadata
- the input list can be empty 
- the items in the input list can be of any type (e.g. str, int, float)
- the linked list items must be in the same order as in the input list 

EXAMPLES 
E1 
input: []
output: None 

E2
input: ["a", "b", "c"]
output: "a" -> "b" -> "c" -> None

E2
input: [1, 2, 3]
output: 1 -> 2 -> 3 -> None

MENTAL MODEL 
Given a list of items, create a linked list containing those values in the list items. 

SCENARIOS
1. iterate over the input list
    - When do stop? 
        - Stop when we reach the end of the list 
        - we could use a for loop
2. attaching the nodes of the linked list in the correct order
    - create `head` variable to instantiate the intial Node 
    - create `tail` variable to guide the next node connection 
    - after we have iterate over the list an created the nodes dynamically, set the tail to None
3. returning the head of the linked list
    - once we break out of the loop, we can return the head 
3. returning None when the input list is empty
    - if the input list is empty, then return None

ALGORITHM 
Parameter List
>> `values`: List[Any]

- IF `values` is empty 
    - RETURN None 

- `head` -> Node instantiated with values[0] value
- `tail` -> Node instantiated with None
- `tail.next` -> `head`

- iterate over `values` from 1 - end of list 
    - `new_node` -> Node instantiated with current value 
    - `tail.next` -> `new_node`
    - `tail` -> `new_node`

- tail.next -> None

- RETURN head

COMPLEXITY 
- N: length of list 
- Time: O(N)
- Space: O(N)
"""

from typing import List, Any

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


# def create_linked_list(values: List[Any]) -> Node:
#     if len(values) == 0:
#         return None
    
#     head = Node(values[0])
#     tail = head

#     for value in values[1:]:
#         new_node = Node(value)
#         tail.next = new_node
#         tail = new_node

#     tail.next = None
#     return head


# def create_linked_list_2(values):
#     dummy_head = Node(None)
#     tail = dummy_head

#     for val in values:
#         tail.next = Node(val)
#         tail = tail.next

#     return dummy_head.next


# time: O(N^2)
# space: O(N)
# def create_linked_list(values):
#     if len(values) == 0:
#         return None
    
#     head = Node(values[0])
#     head.next = create_linked_list(values[1:])  # this creates a new copy of the list 
#     return head

# time: O(N)
# space: O(N)
def create_linked_list(values, i=0):
    if i == len(values):
        return None
    head = Node(values[i])
    head.next = create_linked_list(values, i + 1)
    return head

def print_linked_list(current_node):
    result = []
    while current_node is not None:
        result.append(str(current_node.val))
        current_node = current_node.next

    if len(result) == 0:
        print(None)
        return

    result = ' -> '.join(result)
    print(result)


print_linked_list(create_linked_list(["h", "e", "y"]))
# # h -> e -> y
print_linked_list(create_linked_list([1, 7, 1, 8]))
# # 1 -> 7 -> 1 -> 8
print_linked_list(create_linked_list(["a"]))
# a
print_linked_list(create_linked_list([]))
# None